File: src/engines/calendar/saka-sunda/sakaSundaEngine.js
const MONTHS = [
  { name: 'Kasa', days: 30 },
  { name: 'Karo', days: 31 },
  { name: 'Katiga', days: 30 },
  { name: 'Kapat', days: 31 },
  { name: 'Kalima', days: 30 },
  { name: 'Kanem', days: 31 },
  { name: 'Kapitu', days: 30 },
  { name: 'Kawalu', days: 31 },
  { name: 'Kasanga', days: 30 },
  { name: 'Kadasa', days: 31 },
  { name: 'Hapitlemah', days: 30 },
  { name: 'Hapitkayu', days: 30 },
]

export const SAKA_SUNDA_ANCHOR = {
  year: 1934,
  month: 0,
  day: 1,
  gregorian: new Date(Date.UTC(2011, 11, 22)),
}

const DAY_MS = 24 * 60 * 60 * 1000

export function isWuntuYear(year) {
  if (year % 128 === 0) return false
  return year % 4 === 0
}

export function getYearType(year) {
  return isWuntuYear(year) ? 'WUNTU' : 'WASTU'
}

export function getYearLength(year) {
  return isWuntuYear(year) ? 366 : 365
}

export function getMonthLength(year, monthIndex) {
  if (monthIndex === 11) {
    return isWuntuYear(year) ? 31 : 30
  }

  return MONTHS[monthIndex].days
}

function getDayStartUTC(date) {
  return Date.UTC(
    date.getUTCFullYear(),
    date.getUTCMonth(),
    date.getUTCDate()
  )
}

function getDayOffset(date) {
  return Math.round(
    (
      getDayStartUTC(date)
      - SAKA_SUNDA_ANCHOR.gregorian.getTime()
    ) / DAY_MS
  )
}

function getMonthAndDay(year, dayOfYear) {
  let remaining = dayOfYear

  for (let month = 0; month < MONTHS.length; month += 1) {
    const monthLength = getMonthLength(year, month)

    if (remaining < monthLength) {
      return {
        month,
        day: remaining + 1,
      }
    }

    remaining -= monthLength
  }

  throw new Error(
    `Invalid Saka Sunda day-of-year: ${dayOfYear}`
  )
}

function moveForward(offset) {
  let year = SAKA_SUNDA_ANCHOR.year
  let remaining = offset

  while (remaining >= getYearLength(year)) {
    remaining -= getYearLength(year)
    year += 1
  }

  const { month, day } = getMonthAndDay(year, remaining)

  return {
    year,
    month,
    day,
  }
}

function moveBackward(offset) {
  let year = SAKA_SUNDA_ANCHOR.year - 1
  let remaining = Math.abs(offset) - 1

  while (remaining >= getYearLength(year)) {
    remaining -= getYearLength(year)
    year -= 1
  }

  const dayOfYear =
    getYearLength(year) - 1 - remaining

  const { month, day } = getMonthAndDay(year, dayOfYear)

  return {
    year,
    month,
    day,
  }
}

export function getSakaSundaCalendar(context) {
  const offset = getDayOffset(context.time.instant)

  const result =
    offset >= 0
      ? moveForward(offset)
      : moveBackward(offset)

  return {
    ...result,
    monthName: MONTHS[result.month].name,
    yearType: getYearType(result.year),
    effectiveDate: context.time.instant,
    boundary: 'MIDNIGHT',
  }
}

File: backend/engines/saka_sunda_engine.py
from datetime import date, datetime, timedelta


MONTHS = [
    {"name": "Kasa", "days": 30},
    {"name": "Karo", "days": 31},
    {"name": "Katiga", "days": 30},
    {"name": "Kapat", "days": 31},
    {"name": "Kalima", "days": 30},
    {"name": "Kanem", "days": 31},
    {"name": "Kapitu", "days": 30},
    {"name": "Kawalu", "days": 31},
    {"name": "Kasanga", "days": 30},
    {"name": "Kadasa", "days": 31},
    {"name": "Hapitlemah", "days": 30},
    {"name": "Hapitkayu", "days": 30},
]


ANCHOR_DATE = date(2011, 12, 22)
ANCHOR_YEAR = 1934


def is_wuntu_year(year: int) -> bool:
    if year % 128 == 0:
        return False
    return year % 4 == 0


def get_year_type(year: int) -> str:
    return "WUNTU" if is_wuntu_year(year) else "WASTU"


def get_year_length(year: int) -> int:
    return 366 if is_wuntu_year(year) else 365


def get_month_length(year: int, month_index: int) -> int:
    if month_index == 11:
        return 31 if is_wuntu_year(year) else 30
    return MONTHS[month_index]["days"]


def get_day_offset(target_date: date) -> int:
    return (target_date - ANCHOR_DATE).days


def get_month_and_day(year: int, day_of_year: int) -> tuple[int, int]:
    remaining = day_of_year

    for month_index in range(len(MONTHS)):
        month_length = get_month_length(year, month_index)

        if remaining < month_length:
            return month_index, remaining + 1

        remaining -= month_length

    raise ValueError(
        f"Invalid Saka Sunda day-of-year: {day_of_year}"
    )


def move_forward(offset: int) -> tuple[int, int, int]:
    year = ANCHOR_YEAR
    remaining = offset

    while remaining >= get_year_length(year):
        remaining -= get_year_length(year)
        year += 1

    month_index, day = get_month_and_day(year, remaining)

    return year, month_index, day


def move_backward(offset: int) -> tuple[int, int, int]:
    year = ANCHOR_YEAR - 1
    remaining = abs(offset) - 1

    while remaining >= get_year_length(year):
        remaining -= get_year_length(year)
        year -= 1

    day_of_year = get_year_length(year) - 1 - remaining
    month_index, day = get_month_and_day(year, day_of_year)

    return year, month_index, day


def get_saka_sunda_calendar(value: date | datetime) -> dict:
    if isinstance(value, datetime):
        target_date = value.date()
    elif isinstance(value, date):
        target_date = value
    else:
        raise TypeError("value must be date or datetime")

    offset = get_day_offset(target_date)

    if offset >= 0:
        year, month_index, day = move_forward(offset)
    else:
        year, month_index, day = move_backward(offset)

    return {
        "year": year,
        "month": month_index,
        "day": day,
        "monthName": MONTHS[month_index]["name"],
        "yearType": get_year_type(year),
        "effectiveDate": target_date.isoformat(),
        "boundary": "MIDNIGHT",
        "meta": {
            "engine": "Saka Sunda",
            "phase": "SOURCE_COMPILED",
            "anchorGregorian": ANCHOR_DATE.isoformat(),
            "anchorYear": ANCHOR_YEAR,
        },
    }
